assembly/src/parsers/ast/mod.rs

```diff
@@ -0,0 +1,563 @@
+use crate::MODULE_PATH_DELIM;
+
+use super::{AssemblyError, Token, TokenStream, Vec};
+
+use std::collections::BTreeMap;
+
+mod types;
+use types::{Instruction, Node, Program};
+
+mod io_ops;
+mod stack_ops;
+mod u32_ops;
+
+#[cfg(test)]
+pub mod tests;
+
+// AST Parser that holds internal state to generate correct ASTs.
+#[allow(dead_code)]
+struct Parser {
+    uses: BTreeMap<String, String>,
+}
+
+impl Parser {
+    // CONSTRUCTOR
+    // --------------------------------------------------------------------------------------------
+    /// Returns a new instance of [Parser] that parses Miden programs into a AST.
+    #[allow(dead_code)] 
+    pub fn new() -> Self {
+        Self {
+            uses: BTreeMap::new(),
+        }
+    }
+
+    // PROGRAM COMPILER
+    // --------------------------------------------------------------------------------------------
+
+    // Compiles the provided source into a AST Program. This Program holds information
+    // that can be directly translated into MAST.
+    #[allow(dead_code)] 
+    pub fn compile(&mut self, source: &str) -> Result<Program, AssemblyError> {
```

Comment:
nit: we should probably call this `parse` rather than `compile`.

Also, I think we should add two different methods: `parse_program` which would parse a program as is now, and `parse_module` which could be used to parse a library module.

The differences between the two are:
* Program: cannot have exported procedures but must have a body.
* Module: cannot have a body, but can have exported procedures.

Reply:
Ah got it, so seems like you're proposing we should just have parse_program and parse_module right? since parse == parse_program

Reply:
so both modules and programs can have local procs right?

Reply:
Yes, both module and programs can have local procs (declared via `proc` keyword).

Reply:
fixed